tests/src/testDoocsImage.cc

```diff
@@ -0,0 +1,113 @@
+  Av(2, 1) = 2;
+  Av(3, 1) = 1;
+  std::vector<uint8_t> expectedData = {8, 0, 7, 0, 6, 0, 5, 0, 4, 0, 3, 0, 2, 0, 1, 0};
+
+  MappedDoocsImg A(A0.data(), A0.capacity(), MappedDoocsImg::InitData::No);
+  IMH headerOut;
+  unsigned char* imgData = A.asDoocsImg(&headerOut);
+  BOOST_CHECK(headerOut.aoi_height == (int)h);
+  BOOST_CHECK(headerOut.aoi_width == (int)w);
+  for(unsigned i = 0; i < w * h * 2; i++) {
+    BOOST_CHECK(imgData[i] == expectedData[i]);
+  }
+}
+
+// generate a test image in given byteArray, which should already have required size
+void generateImage(ChimeraTK::OneDRegisterAccessor<uint8_t>& acc) {
+  ChimeraTK::MappedImage im(acc, MappedDoocsImg::InitData::Yes);
+  unsigned w = 20, h = 10;
+  im.setShape(w, h, ImgFormat::Gray16);
+  ChimeraTK::ImgHeader* imh = im.header();
+  imh->effBitsPerPixel = 14;
+  imh->frame = 3;
+
+  auto imv = im.interpretedView<uint16_t>();
+  for(unsigned x = 0; x < w; x++) {
+    unsigned x1 = (x + 5) % w;
+    for(unsigned y = 0; y < h; y++) {
+      imv(x, y) = x1 * x1 - y * y;
+    }
+  }
+}
+
+BOOST_AUTO_TEST_CASE(fromDeviceTest) {
+  std::pair<boost::shared_ptr<ControlSystemPVManager>, boost::shared_ptr<DevicePVManager>> pvManagers =
+      createPVManager();
+  boost::shared_ptr<ControlSystemPVManager> csManager = pvManagers.first;
+  boost::shared_ptr<DevicePVManager> devManager = pvManagers.second;
+
+  static const size_t arraySize = 400000;
+  static const size_t arraySizeCheck = 400;
+  typename boost::shared_ptr<ChimeraTK::NDRegisterAccessor<uint8_t>> deviceVariable =
+      devManager->createProcessArray<uint8_t>(
+          SynchronizationDirection::deviceToControlSystem, "fromDeviceVariable", arraySize);
+  typename boost::shared_ptr<ChimeraTK::NDRegisterAccessor<uint8_t>> controlSystemVariable =
+      csManager->getProcessArray<uint8_t>("fromDeviceVariable");
+
+  DoocsUpdater updater;
+
+  // EqFct = NULL, not required for our purpose
+  DoocsImage doocsImage(nullptr, "someName", controlSystemVariable, updater);
+
+  // generate an image and send it via the device side
+  ChimeraTK::OneDRegisterAccessor acc(deviceVariable);
+  generateImage(acc);
+  deviceVariable->write();
+
+  // everything should still be 0 on the CS side
+  std::vector<uint8_t>& csVector = controlSystemVariable->accessChannel(0);
+  for(size_t i = 0; i < arraySizeCheck; ++i) {
+    BOOST_CHECK(csVector[i] == 0);
+  }
+  // image data not yet set
+  BOOST_CHECK(doocsImage.value() == 0);
+
+  updater.update();
+
+  // The actual vector buffer has changed. We have to get the new reference.
+  csVector = controlSystemVariable->accessChannel(0);
+  auto deviceVector = deviceVariable->accessChannel(0);
+  for(size_t i = 0; i < arraySizeCheck; ++i) {
+    BOOST_CHECK_EQUAL(csVector[i], deviceVector[i]);
+    // note, this works only if the image format is not changed by mapping to DOOCS
+    BOOST_CHECK_EQUAL(doocsImage.value()[i], deviceVector[i + sizeof(ImgHeader)]);
+  }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```

Comment:
The test does not check the creation via the PV factory, the doocs variable config parsing etc, i.e. the whole chain using the DoocsServerTestHelper. I guess the helper would have to be extended to get access to the new type for that.

Reply:
True, it's not an end-to-end test. And as you said, it's another project (DoocsServerTestHelper) that we would need to touch if we want that.

Reply:
It's not crucial for this merge request. Can you create a ticket that the tests should be improved and close this conversation?

Reply:
done: https://redmine.msktools.desy.de/issues/10742